### ISSUE TITLE
 Out of memory in io_realm_internal_SharedGroup.cpp

### ISSUE THREAD

**Issue by ImGauravC** [realm/realm-java#794]
```java
java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 101
       at io.realm.internal.SharedGroup.createNativeWithImplicitTransactions(SharedGroup.java)
       at io.realm.internal.SharedGroup.<init>(SharedGroup.java:55)
       at io.realm.Realm.<init>(Realm.java:160)
       at io.realm.Realm.createAndValidate(Realm.java:435)
       at io.realm.Realm.create(Realm.java:414)
       at io.realm.Realm.getInstance(Realm.java:332)
       at io.realm.Realm.getInstance(Realm.java:294)
       at io.realm.Realm.getInstance(Realm.java:275)

**Comment by cmelchior** [comment]
Hi @ImGauravC 

That error is happening because the device has run out of memory. It can happen in a number of scenarios, the most likely being:

1) You a trying to commit a too large transaction
2) You are not closing your Realms on background threads, and keep making new threads.
3) You have too many open Realms across threads that doesn't auto update, hence are "different" versions.

We really need more information to give an informed answer, but given you get the error when calling .getInstance() I guess that 2) is the most likely culprit.

**Comment by ImGauravC** [comment]
I got this in a UI thread and I access realm instances in just 2 threads  (1 UI and another background) and I properly close the realm instances everytime i open them and it's anyway a thread singleton.

**Comment by sergeyfitis** [comment]
@ImGauravC 1) You a trying to commit a too large transaction. How much items you save per transaction?
Model of your device on which you test?

**Comment by bmunkholm** [comment]
Be aware that you have to call close() for every time you called getInstance().

**Comment by Dexteroo7** [comment]
Would open Realms in a different "process" lead to a memory error because of reason 3) ?

**Comment by cmelchior** [comment]
Realm doesn't support multiprocess access yet, so it would fail. I doubt it would throw a out of memory error though.

**Comment by ImGauravC** [comment]
a) I call close exactly as many times as I call getInstance
b) I'm using Realm in only two thread - Main and a background thread. 
c) What exactly would be a "large transaction"? I'm committing only text messages to the realm.

**Comment by Dexteroo7** [comment]
@ImGauravC  can you check your app's data size in your settings. Also clear your data and try again. I think you are facing same issue as me. In my case write transactions keep increasing the app size.

**Comment by ImGauravC** [comment]
```java
java.lang.RuntimeException: Unable to start activity ComponentInfo{[package_name].MessagingActivity}: java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 101
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2305)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2363)
       at android.app.ActivityThread.access$900(ActivityThread.java:161)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1265)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:157)
       at android.app.ActivityThread.main(ActivityThread.java:5356)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1265)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1081)
       at dalvik.system.NativeStart.main(NativeStart.java)
Caused by: java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 101
       at io.realm.internal.SharedGroup.createNativeWithImplicitTransactions(SharedGroup.java)
       at io.realm.internal.SharedGroup.<init>(SharedGroup.java:55)
       at io.realm.Realm.<init>(Realm.java:159)
       at io.realm.Realm.createAndValidate(Realm.java:441)
       at io.realm.Realm.create(Realm.java:418)
       at io.realm.Realm.getInstance(Realm.java:336)
       at io.realm.Realm.getInstance(Realm.java:298)
       at io.realm.Realm.getInstance(Realm.java:279)
       at [package_name].MessagingActivity.onCreate(MessagingActivity.java:100)
       at android.app.Activity.performCreate(Activity.java:5426)
       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1105)
       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2269)
       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2363)
       at android.app.ActivityThread.access$900(ActivityThread.java:161)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1265)
       at android.os.Handler.dispatchMessage(Handler.java:102)
       at android.os.Looper.loop(Looper.java:157)
       at android.app.ActivityThread.main(ActivityThread.java:5356)
       at java.lang.reflect.Method.invokeNative(Method.java)
       at java.lang.reflect.Method.invoke(Method.java:515)
       at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1265)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1081)
       at dalvik.system.NativeStart.main(NativeStart.java)

**Comment by Dexteroo7** [comment]
@realm-ci  I was getting a similar out of memory error when I tried to rapidly commit transactions on a background thread. Surprisingly the app data size increased exponentially (checked in settings). Adding a change listener to my main thread (UI) realm fixed this. I will share my project as soon as I am able to :)

**Comment by emanuelez** [comment]
Are you having long living transactions in your app? By long living I mean
seconds between beginTransaction() and commitTransaction.

Also, you say you only have two threads, UI and background. How does the
background thread look like? Is it a normal Thread? Does it have a Looper?
Is it an AsyncTask? the more details you can share the better. We really
want to get help you and get to the bottom of this.

On Mon, Feb 2, 2015 at 9:20 PM, Gaurav <notifications@github.com> wrote:

> java.lang.RuntimeException: Unable to start activity ComponentInfo{[package_name].MessagingActivity}: java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 101
>        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2305)
>        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2363)
>        at android.app.ActivityThread.access$900(ActivityThread.java:161)
>        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1265)
>        at android.os.Handler.dispatchMessage(Handler.java:102)
>        at android.os.Looper.loop(Looper.java:157)
>        at android.app.ActivityThread.main(ActivityThread.java:5356)
>        at java.lang.reflect.Method.invokeNative(Method.java)
>        at java.lang.reflect.Method.invoke(Method.java:515)
>        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1265)
>        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1081)
>        at dalvik.system.NativeStart.main(NativeStart.java)Caused by: java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 101
>        at io.realm.internal.SharedGroup.createNativeWithImplicitTransactions(SharedGroup.java)
>        at io.realm.internal.SharedGroup.<init>(SharedGroup.java:55)
>        at io.realm.Realm.<init>(Realm.java:159)
>        at io.realm.Realm.createAndValidate(Realm.java:441)
>        at io.realm.Realm.create(Realm.java:418)
>        at io.realm.Realm.getInstance(Realm.java:336)
>        at io.realm.Realm.getInstance(Realm.java:298)
>        at io.realm.Realm.getInstance(Realm.java:279)
>        at [package_name].MessagingActivity.onCreate(MessagingActivity.java:100)
>        at android.app.Activity.performCreate(Activity.java:5426)
>        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1105)
>        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2269)
>        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2363)
>        at android.app.ActivityThread.access$900(ActivityThread.java:161)
>        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1265)
>        at android.os.Handler.dispatchMessage(Handler.java:102)
>        at android.os.Looper.loop(Looper.java:157)
>        at android.app.ActivityThread.main(ActivityThread.java:5356)
>        at java.lang.reflect.Method.invokeNative(Method.java)
>        at java.lang.reflect.Method.invoke(Method.java:515)
>        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1265)
>        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1081)
>        at dalvik.system.NativeStart.main(NativeStart.java)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/issues/794#issuecomment-72529968>.
>



-- 
*Emanuele Zattin* (@emanuelez <https://twitter.com/emanuelez>)
SW Engineer & CI Specialist

**Comment by ImGauravC** [comment]
a) Yes, I do have long living transactions in my app
b) The background thread is a looper thread. https://github.com/realm/realm-java/issues/726#issuecomment-69439067 , this is the pattern I'm following.

**Comment by emanuelez** [comment]
a) Having long lived transactions is a very bad pattern to follow,
whichever database you are using, not only Realm. I suggest you change that
as a first thing. The longer a transaction is the more memory will be
needed to hold all the versions of your Realm, which can of course lead to
your memory issues.

b) Since it's a Looper thread it's not a good idea to continuously call
getInstance() and close(). You can see how we setup a worker thread in our
adapter example here:
https://github.com/realm/realm-java/blob/master/examples/adapterExample/src/main/java/io/realm/examples/realmadapters/WorkerThread.java
Ad you can see we keep using the same Realm instance and only call close in
the finally block, which will be called when the Looper is killed.


On Mon, Feb 2, 2015 at 9:43 PM, Gaurav <notifications@github.com> wrote:

> a) Yes, I do have long living transactions in my app
> b) The background thread is a looper thread. #726 (comment)
> <https://github.com/realm/realm-java/issues/726#issuecomment-69439067> ,
> this is the pattern I'm following.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/issues/794#issuecomment-72533787>.
>



-- 
*Emanuele Zattin* (@emanuelez <https://twitter.com/emanuelez>)
SW Engineer & CI Specialist

**Comment by ImGauravC** [comment]
I had looked at the example. I continuously call for getInstance() and close() because I couldn't find an easy way to reference the realm instance of the Looper thread in the Runnable that I'll post to it. Can you suggest a way?

**Comment by emanuelez** [comment]
You could send plain messages instead of runnables. You could implement the
logic of the runnable in a Handler instead. See the worker handler in the
same adapter example.

On Mon, Feb 2, 2015 at 10:18 PM, Gaurav <notifications@github.com> wrote:

> I had looked at the example. I continuously call for getInstance() and
> close() because I couldn't find an easy way to reference the realm instance
> of the Looper thread in the Runnable that I'll post to it. Can you suggest
> a way?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/issues/794#issuecomment-72539896>.
>



-- 
*Emanuele Zattin* (@emanuelez <https://twitter.com/emanuelez>)
SW Engineer & CI Specialist

**Comment by ImGauravC** [comment]
Ok. Also what exactly is the issue with calling getInstance and close multiple times? I thought since it's a thread singleton, it won't matter.

**Comment by emanuelez** [comment]
If you call getInstance N times and then close N times, then the only Nth
time you call close will actually perform the cleanup of the native
resources and delete the cached Realm instance. But since you keep calling
getInstance once and close once (but several times in a row), then every
call to close will perform the cleanup. Also getInstance will have to
instantiate from scratch which is a fairly expensive operation.

On Mon, Feb 2, 2015 at 10:52 PM, Gaurav <notifications@github.com> wrote:

> Ok. Also what exactly is the issue with calling getInstance and close
> multiple times? I thought since it's a thread singleton, it won't matter.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/issues/794#issuecomment-72545859>.
>



-- 
*Emanuele Zattin* (@emanuelez <https://twitter.com/emanuelez>)
SW Engineer & CI Specialist

**Comment by ImGauravC** [comment]
Also how long a transaction would be safe? Currently the longest transaction I make takes around 300ms.

**Comment by creativepsyco** [comment]
**TL;DR**

I have a total of 3 Realm instances in the entire application and I still have this issue. My local db size is about 100 MBs. From time to time I get this stack trace:

```
Realm Version: 0.78.0
02-03 23:14:26.367    6957-6957/com.example.android E/AndroidRuntime﹕ FATAL EXCEPTION: main
    Process: com.example.android, PID: 6957
    java.lang.RuntimeException: Unspecified exception. mremap(): failed: Out of memory in io_realm_internal_SharedGroup.cpp line 144
            at io.realm.internal.SharedGroup.nativeAdvanceRead(Native Method)
            at io.realm.internal.SharedGroup.advanceRead(SharedGroup.java:88)
            at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:30)
            at io.realm.Realm$RealmCallback.handleMessage(Realm.java:216)
            at android.os.Handler.dispatchMessage(Handler.java:98)
            at android.os.Looper.loop(Looper.java:146)
            at android.app.ActivityThread.main(ActivityThread.java:5748)
            at java.lang.reflect.Method.invokeNative(Native Method)
            at java.lang.reflect.Method.invoke(Method.java:515)
            at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1291)
            at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1107)
            at dalvik.system.NativeStart.main(Native Method)
```

Its kind of becoming pretty unusable once the DB size grows beyond 100 MBs. I have 4-5 tables with some have relationships among them. When the DB size is small I had no issues, reading or writing data. For writing data I am using the pattern as suggested in the `WorkerHandler` example in the docs. After a while, making a `findAll` call for reading also causes Out of memory issues. Can 3 realms cause this much damage? I have hunted down Java related issues and have run out of clues. Its upto the native part now, which unfortunately I can't debug.

**Comment by emanuelez** [comment]
Hello Mohit.
Do you have long living transactions? Do you have different Realm instances
at different versions for long periods of time?
Can you by any chance share your code, even in a private way?

On Tue, Feb 3, 2015 at 4:31 PM, Mohit Kanwal <notifications@github.com>
wrote:

> *TL;DR*
>
> I have a total of 3 Realm instances in the entire application and I still
> have this issue. My local db size is about 100 MBs. From time to time I get
> this stack trace:
>
> Realm Version: 0.78.0
> 02-03 23:14:26.367    6957-6957/com.example.android E/AndroidRuntime﹕ FATAL EXCEPTION: main
>     Process: com.example.android, PID: 6957
>     java.lang.RuntimeException: Unspecified exception. mremap(): failed: Out of memory in io_realm_internal_SharedGroup.cpp line 144
>             at io.realm.internal.SharedGroup.nativeAdvanceRead(Native Method)
>             at io.realm.internal.SharedGroup.advanceRead(SharedGroup.java:88)
>             at io.realm.internal.ImplicitTransaction.advanceRead(ImplicitTransaction.java:30)
>             at io.realm.Realm$RealmCallback.handleMessage(Realm.java:216)
>             at android.os.Handler.dispatchMessage(Handler.java:98)
>             at android.os.Looper.loop(Looper.java:146)
>             at android.app.ActivityThread.main(ActivityThread.java:5748)
>             at java.lang.reflect.Method.invokeNative(Native Method)
>             at java.lang.reflect.Method.invoke(Method.java:515)
>             at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1291)
>             at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1107)
>             at dalvik.system.NativeStart.main(Native Method)
>
> Its kind of becoming pretty unusable once the DB size grows beyond 100
> MBs. I have 4-5 tables with some have relationships among them. When the DB
> size is small I had no issues, reading or writing data. For writing data I
> am using the pattern as suggested in the WorkerHandler example in the
> docs. After a while, making a findAll call for reading also causes Out of
> memory issues. Can 3 realms cause this much damage? I have hunted down Java
> related issues and have run out of clues. Its upto the native part now,
> which unfortunately I can't debug.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/realm/realm-java/issues/794#issuecomment-72670553>.
>



-- 
*Emanuele Zattin* (@emanuelez <https://twitter.com/emanuelez>)
SW Engineer & CI Specialist

**Comment by creativepsyco** [comment]
Hi Emanuele,
Here's a gist of the necessary guilty files: https://gist.github.com/creativepsyco/f7245f0e3cd803e573c8

I have 3 places in the app where I make write transactions from, these are 
* The TCP socket connection with the server: This is taken care of by the Worker Thread & Handler Thread to which I post the updates, implementation basically mimics the Realm example 
* The RESTful API Responses: This is handled by callbacks which happen on a single dedicated thread
* The SyncAdapter: It periodically fetches data, Its an implementation of the AbstractThreadedSyncAdapter running on a single thread. However, at the end of its update cycle I make sure to call `realm.close()`

On top of that any references to query making use of the Realm instance need to pass via DBManager which maintains a cache of Realm instances. It will lookup the thread info and create the Realm instances if necessary or give it a cached realm instance based on the thread id. This way I am able to point out any pieces of code making use of closed Realm instances. 

I guess there is a bit of documentation lacking regarding any established patterns of use especially in a Multi-threaded application. Let me know if you need more code sample. I guess I can move all writes to a single ThreadHandler however, I still need the query part across multiple threads, AsyncTasks etc. which still need their own Realm Instances, eventually closing them down when they are done.

**Comment by emanuelez** [comment]
Thanks for the code! I will take a look at it. But I can say one thing immediately: you don't need to do the Realm instance caching because we already do :) I'll be back after looking at your code.

**Comment by emanuelez** [comment]
I think the reason for your OOM problem is the long lived transaction. two seconds is an eternity in Realm land. My suggestion is to first gather your data from REST and only then actually insert it in the Realm. Such long lived transactions force Realm to have to maintain many versions of the data, inflating the memory consumption.

**Comment by ImGauravC** [comment]
With Mohit's design, we won't have to worry about calling closeable everytime we do a Realm.getInstance, right? Isn't that better?

**Comment by emanuelez** [comment]
It's not better because DBManager is not able to know when it's time to call the `close()` method. The risk is to call close() too early and get all your objects and results invalidated.

**Comment by creativepsyco** [comment]
> I think the reason for your OOM problem is the long lived transaction. two seconds is an eternity in Realm land.
Wow, this, I didn't know and I shall optimise further. I was kinda hoping to do transactions in batch but seems like I should be putting my `beginTransaction()` and `commitTransaction()` within the loop itself instead of outside.

I cached the instances in the DBManager because I wanted to debug which instances are being hogged by which threads, dump their ids, state (whether closed/open etc), this is mostly a proxy call, it doesn't call `close()` or any other lifecycle methods because it cannot know when to do so.

**Comment by denley** [comment]
I am encountering this same error in one of my projects. It hasn't happened on my test device, but I am getting some crash reports coming in for an app I've released (I know, Realm is still in beta so I probably shouldn't be using it in production. But I did). Some of them say `Out of memory in io_realm_internal_SharedGroup.cpp line 101` and some say `Too many open files in io_realm_internal_SharedGroup.cpp` (all happening on a call to `Realm.getInstance`. This crash occurs in a native method `io.realm.internal.SharedGroup.createNativeWithImplicitTransactions`.

Mine is happening inside a call to `onHandleIntent` of an `IntentService` that receives GPS location updates every three seconds, so the thread has a looper and the realms are not being used concurrently (An `IntentService` processes only one intent at a time. But in any case I doubt my method is taking three seconds to run). I'm not making any network connections, or other long processes. Each time it makes one transaction (one object creation and one object update) and a some reads too.

I'm using realm version 0.78.0.

**Comment by denley** [comment]
From the comments above I'm guessing that my problem is that I'm creating and closing Realms too quickly (once every three seconds). 

This is an annoying limitation since it's the only way to work within the framework of an `IntentService` (the `Thread` and `Handler` are all private). I guess that I will have to use a `Service` instead and construct my own handler thread and realm singleton (effectively re-implement the `IntentService` class).

It would be great if there were a solution to make this kind of implementation simpler. Perhaps by caching realm instances until the process dies, or maybe by packaging a `RealmIntentService` class.

**Comment by cmelchior** [comment]
Hi @denley 
Using a IntentService should be OK. If you are getting to many files open, you are hitting the file descriptor limit in the operating system. That number is usually above 1000, and would indicate that Realms are not being closed. Are you by any chance spawning your own threads?

**Comment by denley** [comment]
thanks for the response @cmelchior 

I've just double checked all of my realms and they are definitely all being closed, and I'm not making any threads in the `IntentService`. Though I have noticed another part of the app that is making a network connection inside a realm transaction (oops), so I guess that could be the culprit, though it should be rare that it is occurring at the same time as the location updates. I'll try fixing that and seeing if the reports keep coming in. Thanks again.

**Comment by cmelchior** [comment]
This seems to have been answered. Feel free to add additional information if this is not the case.

**Comment by timanglade** [comment]
Reopening this issue since many users seem to be hitting this error message for many different reasons. Maybe we should output a nicer, more helpful error message here, along with a pointer to the docs about likely causes and remedies? (e.g. /docs/java/#out-of-memory-errors). Otherwise people may think this is a) a bug in Realm b) something we haven’t fixed yet and c) keep adding comment to this issue every time they get the error message. Thoughts @cmelchior @emanuelez @kneth ?

**Comment by bmunkholm** [comment]
Created separate issue for improved help.

**Comment by idelgado** [comment]
I am using realm 0.79.1 with encryption enabled on Android. I am copying the data contents of a file into realm by splitting the file into 1MB chunks. The files can be up to 250MB.

The thread reads the data using RandomAccessFile.read() and copies the data to realm. The loop runs about every 100ms and calls copyDataToRealm. After about 110MB of data (110 iterations of the loop) I will start getting out of memory exception below.

My impression is that the chunk size and database size has something to do with the out of memory exception. The reason I say this is because I can store the same data across multiple realm instances(separate underlying realm files) with no problem if I limit the number of chunks to 50, or a max size of 50MB of file data per database. Otherwise the database write transaction throws an out of memory exception after about 110MB. If I try to use the same database later to write anything, lets say 100K of data I also get an out of memory exception.

I noticed a comment above that also described a similar behavior once they got over 100MB. Not sure if that is a coincidence or a pattern but figured i'd share as well.

Do you run any instrumentation tests similar to this kind of situation? Large blobs resulting in large database sizes. If so could you share these tests so I can run them on my device.

```java
    private void copyDataToRealm(String name, byte[] data, int seq, String realmDataStorage) {
        Realm realm = null;
        try {
            realm = Realm.getInstance(appContext, realmDataStorage, key);
            realm.beginTransaction();
            RealmData dataRealm = realm.createObject(RealmData.class);
            dataRealm.setName(name);
            dataRealm.setData(data);
            dataRealm.setSeq(seq);
            realm.commitTransaction();
        } finally {
            if(realm != null) {
                realm.close();
            }
        }
    }
```

```
3-05 13:45:30.609    D/REALM﹕ jni: ThrowingException 11, mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 164, .
03-05 13:45:30.609    D/REALM﹕ Exception has been throw: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 164
03-05 13:45:30.750    W/System.err﹕ java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 164
03-05 13:45:30.750    W/System.err﹕ at io.realm.internal.SharedGroup.nativeCommitAndContinueAsRead(Native Method)
03-05 13:45:30.750    W/System.err﹕ at io.realm.internal.SharedGroup.commitAndContinueAsRead(SharedGroup.java:103)
03-05 13:45:30.750    W/System.err﹕ at io.realm.internal.ImplicitTransaction.commitAndContinueAsRead(ImplicitTransaction.java:45)
03-05 13:45:30.750    W/System.err﹕ at io.realm.Realm.commitTransaction(Realm.java:1490)
03-05 13:45:30.750    W/System.err﹕ at io.pristine.runnable.StreamDataToRealmRunnable.copyDataToRealm(StreamDataToRealmRunnable.java:184)
```

**Comment by kneth** [comment]
Realm is opening data files using `mmap()` at the C/C++ level. That implies that a realm is not directly allocated on the JVM heap, and you can open much larger data sets.

But of course, no devices have unlimited amount of memory, and from your description it sounds like that your app is using more than the device's physical memory (many Android devices don't have swap space).

It is very hard to prevent such an out-of-memory error. But we will always leave the realm file consistent so you don't loose already saved/committed data.

**Comment by kesslermichael** [comment]
Hello, we face a similar issue in our project.

Please refer to the next SO question (Christian has already answered there):
http://stackoverflow.com/questions/29396200/how-to-use-realmbaseadapter-without-memory-exceptions/29420175 

Thank you.

**Comment by cmelchior** [comment]
Hi @kesslermichael 
Are you able to share your code? If you want to keep it private you can send info to help@realm.io. Or create a reproducible example?

**Comment by kesslermichael** [comment]
Hi Christian, 
We work in a reproducible example. 
I believe tomorrow we'll be able to post it here. 
Thank you.

**Comment by cmelchior** [comment]
Hi @kesslermichael 
Have you been able to reproduce this?

**Comment by kesslermichael** [comment]
Hi, 

Sorry for not posting back. 
We have found that the OOM crash happens when trying to call getInstance on the Realm thread (thread that we manage for performing all the write transactions to realm) exactly in the same time when another thread performs a query to the contacts of the device. 

We have walked through tens of OOM crashes (in Crashlytics) and have come to this conclusion. 
It is much more frequent on devices with many contacts (more than 1K). 

Currently we have changed the implementation of contacts fetching so that the chance that these 2 calls will happen in the same time will be minimal and seems that the OOM doesn't happen anymore. 
I hope that it will be the same in production. 
Until now we have updated each contact in its own transaction. 
Now we collect one huge JSON and call the CreateOrUpdateAllFromJSON method after all the queries to contacts are completed. 

Seems like the query to contact uses too many memory - as the getInstance (when there is no open realm in the calling thread). 
But I suggest you to try to reproduce it and see if it may somehow be solved on your side. 

Additional background: 
In the realm thread we open and close the realm for each transaction - not holding one open instance for the entire app life (this way some other issue was solved for us).

Thanks.

**Comment by eboudrant** [comment]
I was able to reproduce it, I will send you a sample app. For each getInstance I close the realm and the transaction are just for 1 entity insertion. I got 

```
                 REALM  D  jni: ThrowingException 11, mremap(): failed: Out of memory in io_realm_internal_SharedGroup.cpp line 164, .
                        D  Exception has been throw: Unspecified exception. mremap(): failed: Out of memory in io_realm_internal_SharedGroup.cpp line 164
        AndroidRuntime  E  FATAL EXCEPTION: MyThread
                        E  Process: sample, PID: 6706
                        E  java.lang.RuntimeException: Unspecified exception. mremap(): failed: Out of memory in io_realm_internal_SharedGroup.cpp line 164
                        E      at io.realm.internal.SharedGroup.nativeCommitAndContinueAsRead(Native Method)
                        E      at io.realm.internal.SharedGroup.commitAndContinueAsRead(SharedGroup.java:103)
                        E      at io.realm.internal.ImplicitTransaction.commitAndContinueAsRead(ImplicitTransaction.java:45)
                        E      at io.realm.Realm.commitTransaction(Realm.java:1469)
                        E      at sampleapp.SampleActivity$2.onLooperPrepared(ChatActivity.java:136)
                        E      at android.os.HandlerThread.run(HandlerThread.java:60)
```
Once crashed each launch crash on : 

```
                 REALM  D  jni: ThrowingException 11, mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 100, .
                        D  Exception has been throw: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 100
        AndroidRuntime  E  FATAL EXCEPTION: MyThread
                        E  Process: sample, PID: 14056
                        E  java.lang.RuntimeException: Unspecified exception. mmap() failed: Out of memory in io_realm_internal_SharedGroup.cpp line 100
                        E      at io.realm.internal.SharedGroup.createNativeWithImplicitTransactions(Native Method)
                        E      at io.realm.internal.SharedGroup.<init>(SharedGroup.java:58)
                        E      at io.realm.Realm.<init>(Realm.java:204)
                        E      at io.realm.Realm.createAndValidate(Realm.java:513)
                        E      at io.realm.Realm.create(Realm.java:486)
                        E      at io.realm.Realm.getInstance(Realm.java:404)
                        E      at io.realm.Realm.getInstance(Realm.java:366)
                        E      at sample.SampleActivity$2.onLooperPrepared(ChatActivity.java:124)
                        E      at android.os.HandlerThread.run(HandlerThread.java:60)
```

I am using ```io.realm:realm-android:0.80.1```

**Comment by kesslermichael** [comment]
Hi @cmelchior,

As I have written, we have decreased the amount of possible memory-heavy operations that might happen in the same time with opening a realm.
But there are still OOM crashes.
It was reduced to less than 1% of sessions (according to Crashlytics) but we still talk about hundreds of crashes a day...

Currently the background thread (the one that is used for all the write transaction on realm) creates a realm object and closes it for each write transaction (calls 'getInstance' before the transaction and 'close' - after).
We have tried the second approach (hold the realm instance for the entire app life) but I think that it caused other problems - I think that it caused OOM as well in some later stage.

We have a couple of possible solutions for this:
1. Surround the 'getInstance' with try-catch and retry getting the instance in catch.
2. Create the realm instance once if few transactions - let's say close it after each 10-th transaction and create before the next transaction. 
3. Try keeping the realm instance in BG thread for the entire app live.

What do you suggest?

**Comment by cmelchior** [comment]
Hi @kesslermichael 
Sorry for the late reply. We actually just fixed a memory leak today in RealmBaseAdapter that might be the source of your problems. You can see the details in #1109 and #1115.

Our CI is having some trouble so I am not sure if you can use our SNAPSHOT release just yet, but building from master should work.

If this is not the cause, I might be stating the obvious, but have you tried looking for memory leaks in general using the Memory Monitor?

If nether of these things help, it would be really helpfull if we could see your source code. OOM errors are notoriously difficult to debug without the actual code unfortunately :(

**Comment by cmelchior** [comment]
Closing this due to no response and that we have implemented a likely fix. Please reopen if you are still seeing problems with this.
